### PR TITLE
GEOMESA-1970 Lazily evaluated simple feature deserialization

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Added `Lazy` option to SerializationOptions
* KryoFeatureSerializer now instantiated with `apply` instead of `new`
* Changed order of ScalaSimpleFeature constructor arguments

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Comment by jahhulbert-ccri** [comment]
I like the change to move the SFT to the first arg, but it will definitely cause some downstream refactoring :)

**Comment by elahrvivaz** [comment]
I added a secondary constructor to preserve the old arg-order. I didn't exhaustively do all the options, b/c scala was complaining about default arguments being used too many times, but I can add them if people are actually using them.

**Comment by jahhulbert-ccri** [comment]
cool :) I'll try to dig through soon.

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/AbstractSimpleFeature.scala [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+  override def getAttribute(name: Name): AnyRef = getAttribute(name.getLocalPart)
+  override def getAttribute(name: String): AnyRef = {
+    val index = sft.indexOf(name)
+    if (index == -1) { null } else { getAttribute(index) }
+  }
+
+  override def getAttributeCount: Int = sft.getAttributeCount
+  override def getAttributes: java.util.List[Object] = {
+    val list = new java.util.ArrayList[AnyRef]
```
Comment (jnh5y) [inline geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/AbstractSimpleFeature.scala]: Is it worth initializing this as an array of length getAttributeCount?

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala [hunk 2]
```diff
@@ -10,169 +10,40 @@ package org.locationtech.geomesa.features
...
-  override def getName: Name = sft.getName
-  override def getUserData: java.util.Map[AnyRef, AnyRef] = userData
+    extends AbstractMutableSimpleFeature(sft, initialId, initialUserData) {
 
-  override def getAttribute(name: Name): AnyRef = getAttribute(name.getLocalPart)
-  override def getAttribute(name: String): AnyRef = {
-    val index = sft.indexOf(name)
-    if (index == -1) null else getAttribute(index)
+  @deprecated("use primary constructor")
+  def this(initialId: String, sft: SimpleFeatureType) = {
```
Comment (jnh5y) [inline geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala]: Ah, nice!  I was worried that all kinds of client code was going to have to change.

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala [hunk 3]
```diff
@@ -10,169 +10,40 @@ package org.locationtech.geomesa.features
...
+  override def getAttribute(index: Int): AnyRef = values(index)
 }
 
 object ScalaSimpleFeature {
 
   def copy(in: SimpleFeature): ScalaSimpleFeature = copy(in.getFeatureType, in)
 
   def copy(sft: SimpleFeatureType, in: SimpleFeature): ScalaSimpleFeature =
-    new ScalaSimpleFeature(in.getID, sft, in.getAttributes.toArray, in.getUserData)
+    new ScalaSimpleFeature(sft, in.getID, in.getAttributes.toArray, new java.util.HashMap[AnyRef, AnyRef](in.getUserData))
```
Comment (jnh5y) [inline geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala]: Does the HashMap constructor do a copy?

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/AbstractSimpleFeature.scala (hunk 1 above)
Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/AbstractSimpleFeature.scala]: yeah good point

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala (hunk 3 above)
Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala]: yeah it basically does `in.foreach(this.put)`

File: geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/AbstractSimpleFeature.scala (hunk 1 above)
Comment (jnh5y) [inline geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/AbstractSimpleFeature.scala]: I kinda assume you leave a few quickies like this to see if I actually read your PRs....

Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/AbstractSimpleFeature.scala]: hahaha i wish

**Pull request merged**